libzfs_core/_nvlist.py

```diff
@@ -0,0 +1,257 @@
+# Copyright 2015 ClusterHQ. See LICENSE file for details.
+
+"""
+nvlist_in and nvlist_out provide support for converting between
+a dictionary on the Python side and an nvlist_t on the C side
+with the automatic memory management for C memory allocations.
+
+nvlist_in and nvlist_out are to be used with the 'with' statement.
+
+nvlist_in takes a dictionary and produces a CData object corresponding
+to a C nvlist_t pointer suitable for passing as an input parameter.
+The nvlist_t is populated based on the dictionary.
+
+nvlist_out takes a dictionary and produces a CData object corresponding
+to a C nvlist_t pointer to pointer suitable for passing as an output parameter.
+Upon exit from a with-block the dictionary is populated based on the nvlist_t.
+
+The dictionary must follow a certain format to be convertible
+to the nvlist_t.  The dictionary produced from the nvlist_t
+will follow the same format.
+
+Format:
+- keys are always byte strings
+- a value can be None in which case it represents boolean truth by its mere presence
+- a value can be a bool
+- a value can be a byte string
+- a value can be an integer
+- a value can be a CFFI CData object representing one of the following C types:
+    int8_t, uint8_t, int16_t, uint16_t, int32_t, uint32_t, int64_t, uint64_t, boolean_t, uchar_t
+- a value can be a dictionary that recursively adheres to this format
+- a value can be a list of bools, byte strings, integers or CData objects of types specified above
+- a value can be a list of dictionaries that adhere to this format
+- all elements of a list value must be of the same type
+"""
+
+import numbers
+from collections import namedtuple
+from contextlib import contextmanager
+from .bindings import libnvpair
+from .ctypes import _type_to_suffix
+
+_ffi = libnvpair.ffi
+_lib = libnvpair.lib
+
+
+@contextmanager
+def nvlist_in(props):
+    """
+    A context manager that converts a python dictionary to a C nvlist_t
+    and provides automatic memory management for the latter.
+    An FFI CData object representing the nvlist_t pointer can be acccessed
+    via 'as' target.
+    """
+    nvlistp = _ffi.new("nvlist_t **")
+    res = _lib.nvlist_alloc(nvlistp, 1, 0) # UNIQUE_NAME == 1
+    if res != 0:
+        raise MemoryError('nvlist_alloc failed')
+    nvlist = _ffi.gc(nvlistp[0], _lib.nvlist_free)
+    _dict_to_nvlist(props, nvlist)
+    # NB 'yield' is the last statement, so this function could be
+    # just a regular function, it is kept as a context manager
+    # only for symmetry with nvlist_out
+    yield nvlist
```

Comment:
> I never explicitly intended for the C nvlist_t object to be usable outside of with scope. At least, I can't think of any use case at the moment.

Indeed.  And only using it inside the scope seems reasonable to me.  And I see that this is how it is used in all cases in this PR.  But the fact that Python leaves it in scope and makes it possible to use it later suggests to me that the context manager is not the way to go because it gives the user the opportunity to screw up really badly whereas the normal function route protects them.